### PR TITLE
Fix stream_ordered_memory_resource attempt to record event in stream from another device

### LINKED ISSUE #1306: [QST] Is get_event()'s thread_local problematic?
https://github.com/rapidsai/rmm/blob/0a5b86cf9958dc195fb70e63e83095f0aa8b33c5/include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp#L288

```c++
stream_event_pair get_event(cuda_stream_view stream)
{
    if (stream.is_per_thread_default()) {
      // Create a thread-local shared event wrapper. Shared pointers in the thread and in each MR
      // instance ensures it is destroyed cleaned up only after all are finished with it.
      thread_local auto event_tls = std::make_shared<event_wrapper>();
      default_stream_events.insert(event_tls);
      return stream_event_pair{stream.value(), event_tls->event};
    }
...
```

For the above code, the thread_local will make the `event_tls` a thread static varible. I think the `event_tls` will be invalid if the thread switches the device it's working on.

Thinking of the code below, I got an assertion failure caused by `cudaEventRecord` when destructing `b`. According to [Stream synchronization behavior](https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html#stream-sync-behavior), "per-thread default stream is an implicit stream local to both the thread and the CUcontext", so I assume the two `cuda_stream_per_thread` should refer to different streams on each device, please correct me if it's wrong.

BTW, it doesn't matter whether it's in PTDS build or not, as long as I'm using `cuda_stream_per_thread`, the assertion failure occurs.

```c++
// initializing pool_memory_resource of multiple devices
...
{
    cudaSetDevice(0);
    rmm::device_uvector<int> a(10, rmm::cuda_stream_per_thread);

    {
        cudaSetDevice(1);
        rmm::device_uvector<int> b(10, rmm::cuda_stream_per_thread);
    }

    cudaSetDevice(0);
}
```

Is this a bug or an intended behavior. If it's an intended behavior, then is it true that in general I shouldn't let the thread switch devices when working with rmm, though I didn't see any  documented restriction about that, maybe I missed something.

### PR DESCRIPTION
**Pull request by harrism**
## Description
As described in #1306 , stream_ordered_memory_resource maintained thread-local storage of an event to synchronize in the per-thread default stream. However, this caused the event to be recorded on the wrong stream if allocations were made on memory resources associated with different devices, because allocation on the first device on the PTDS would initialize the TLS for that stream, and subsequent device pools would try to use the already initialized TLS. 

This PR adds a new test that only runs on multidevice systems (more correctly, does nothing on single device systems). The test creates two per-device pools, and creates and destroys a device buffer on each. 

It also fixes `stream_ordered_memory_resource` to store the ID of the device that is current when it is constructed, and then to store a vector of one event per device in TLS rather than a single event. When the PTDS is passed to `get_event`, the event for the MR's stored device ID is used. This should correctly support PTDS with multiple threads and multiple devices (still one MR per device).

The PR also includes some changes to the device ID utilities in `cuda_device.hpp`. There is a new RAII device helper class, and a `get_num_cuda_devices()` function.

Finally, there is a small addition to the .clang-tidy to disable warnings about `do...while` loops inside of RMM error checking macros.

- Fixes #1306 

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (wence-) [inline include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:300]: So this creates an event on the currently active device, so there's an implicit requirement that `get_event` is called with `cudaGetDevice` equal to `this->device_id_`. See discussion below on whether we want to ensure this internally, or document the requirements explicitly.

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (wence-) [inline include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:313]: Likewise here, the event is created on whichever device is currently active (which may not be `this->device_id_`).

**Review by wence- — CHANGES_REQUESTED** [review]
Thanks @harrism. These changes look correct to me, I mostly have an additional query about what guarantees we want RMM to be providing in a multi-device setting wrt user management of device context switching.

Rationale: I still think there are circumstances which one might _expect_ to work but which do not with these changes, and I suppose we should decide if we want to make those work, or document that they do not, and why.

File: tests/mr/device/pool_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/pool_mr_tests.cpp:184]: If we do that, if it fails, it fails with an assert, which causes the test to exit. This happens to be the last test in the binary, so all tests will finish, but the gtest summary will not be printed.

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:300]: I think we do document that a memory resource is only valid to use on the device where it was created. That means the device must be current before calling any memory resource methods. Next week I will seek out the documentation and make sure it's sufficient.

File: tests/mr/device/pool_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/pool_mr_tests.cpp:187]: I'd like to spend some time thinking about this. We may want our RAII data structures to ensure they activate the device they were allocated on in their destructors...

Or we may not. That may forcing performance costs on the user under the covers when they might not need the help.

File: tests/mr/device/pool_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/pool_mr_tests.cpp:180]: Good question. The MRs are thread safe. Moreover, in the CUDA runtime, the current device is maintained per thread. In other words cudaSetDevice sets the current device for the calling host thread. So you don't need to worry about critical sections for current device management. 

See https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g159587909ffa0791bbe4b40187a4c6bb

File: tests/mr/device/pool_mr_tests.cpp
Comment (wence-) [reply tests/mr/device/pool_mr_tests.cpp:184]: Ah, ok.

File: tests/mr/device/pool_mr_tests.cpp
Comment (wence-) [reply tests/mr/device/pool_mr_tests.cpp:180]: Thanks!

File: tests/mr/device/pool_mr_tests.cpp
Comment (wence-) [reply tests/mr/device/pool_mr_tests.cpp:187]: Thanks. Given your comment below about UB in the docs, I've updated my comment to answer some of the queries: we don't need to handle device switching in allocation, and my usage example is invalid.

Comment (wence-) [reply tests/mr/device/pool_mr_tests.cpp:187]: > Or we may not. That may forcing performance costs on the user under the covers when they might not need the help.

Yeah, I am not sure which side of then fence I am on here.

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (wence-) [reply include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:313]: As noted, if the active device doesn't match `device_id_` this is UB per the docs.

**Comment by wence-** [comment]
The cpp builds failed due to (I think) https://github.com/conda-forge/rhash-feedstock/issues/25. It looks like an updated cmake conda package is now built so I have rerun the tests :crossed_fingers:

File: tests/mr/device/pool_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/pool_mr_tests.cpp:187]: After thinking about this, and discussing with @jrhemstad , I am inclined to leave this as UB for now. I don't think we want to be switching devices back and forth every time we `mr->deallocate()`. I'm not confident that this can be done without a large impact on performance, and there will be situations where the user has more information and can optimize device switching. 

If it becomes a problem for containers, perhaps we can add it at the `allocator` level, rather than the `memory_resource` level.

Comment (wence-) [reply tests/mr/device/pool_mr_tests.cpp:187]: Thanks, this makes sense. 

Suggestion (non-blocking): Shall we add a sentence explicitly mentioning destructors to the "Multiple devices" section of the documentation.

Perhaps after the existing first paragraph:

> A device_memory_resource should only be used when the active CUDA device is the same device that was active when the device_memory_resource was created. Otherwise behavior is undefined.

"Note that this means that destructors must run with the same CUDA device active as used when creating their associated memory resource."

WDYT?

**Review by wence- — APPROVED** [review]

**Review by wence- — APPROVED** [review]
Thanks @harrism, I've dismissed my requested changes, and propose one non-blocking suggestion to add an explicit sentence about destructors in the multiple-device setting to the docs. (My review has no power here though, since not a cpp-codeowner...)

**Comment by harrism** [comment]
@wence- Thank you for your attention to detail! Even if you are not a cpp-codeowner, since you reviewed so carefully, I count your approval toward the 2 C++ approvals we require.

File: tests/mr/device/pool_mr_tests.cpp
Comment (harrism) [reply tests/mr/device/pool_mr_tests.cpp:187]: Yes. Let me add something along these lines.

Comment (harrism) [reply tests/mr/device/pool_mr_tests.cpp:187]: I added doc and example.

**Review by bdice — APPROVED** [review]
I reviewed the issue, code, and prior PR discussions here. To the best of my knowledge, this looks like a good fix. I learned a fair bit while reviewing this PR, so thanks for the lessons @harrism @wence- !

File: README.md
Comment (harrism) [reply README.md:369]: Good catch! Thank you.

File: include/rmm/cuda_device.hpp
Comment (jrhemstad) [inline include/rmm/cuda_device.hpp:75]: **note (non-blocking):** In CUB's version of the equivalent type, it looks like it attempts to avoid an extraneous `cudaSetDevice` call in the dtor if the new device was the same as the old: https://github.com/NVIDIA/cccl/blob/f2b931bfddd1d06bd96ca9d65506625144256f78/cub/cub/util_device.cuh#L92-L110

No idea if this was just premature optimization or actually useful.

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (jrhemstad) [inline include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp:302]: **suggestion:** There's an extraneous `std::shared_ptr` copy here. This is one way to eliminate it
```suggestion
        events_tls[device_id.value()] = event;
        this->default_stream_events.emplace(std::move(event));
```

**Review by jrhemstad — APPROVED** [review]
The extra `shared_ptr` copy isn't the end of the world, so you could always address that later.

File: include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp]: Fixed.

File: include/rmm/cuda_device.hpp
Comment (harrism) [reply include/rmm/cuda_device.hpp:75]: Fixed, thanks.

**Pull request merged**
